### PR TITLE
Scaling and Multigrid

### PR DESCRIPTION
**Pull request by christophermaynard**
Scaling and Multigrid sections are now written. They need a reviewer!

### CODE REVIEW

File: multigrid.tex [hunk 1]
```diff
@@ -1 +1,93 @@
...
+PSyclone. These have all been successfully implemented. Thirdly, when
+the multigrid algorithm itself was implemented the solver had problems
+converging.
+This was ultimately resolved as an issue with the vertical
+pre-conditioner and not multigrid itself. 
+
+The multigrid algorithm implemented here is a four-level V-cycle of
+the horizontal mesh, and used to pre-condition the Helmholtz or
+pressure solver. In fact, the pre-conditioned pressure field is taken
+as the ``solved'' outcome and no iterative, Krylov solver is applied
```
Comment (samcusworth) [inline multigrid.tex]: "no iterative" -> "non-iterative"?

File: multigrid.tex [hunk 2]
```diff
@@ -1 +1,93 @@
...
+implementation. Secondly, implementing multigrid itself required
+several infrastructure changes. Multiple meshes, maps between these
+meshes and support for intergrid kernels in the PSyKAl API and in
+PSyclone. These have all been successfully implemented. Thirdly, when
+the multigrid algorithm itself was implemented the solver had problems
+converging.
+This was ultimately resolved as an issue with the vertical
+pre-conditioner and not multigrid itself. 
+
+The multigrid algorithm implemented here is a four-level V-cycle of
```
Comment (samcusworth) [inline multigrid.tex:17]: It's not really pertinent here but it's bugging me: Why (only) 4 levels?

File: multigrid.tex [hunk 3]
```diff
@@ -1 +1,93 @@
...
+the horizontal mesh, and used to pre-condition the Helmholtz or
+pressure solver. In fact, the pre-conditioned pressure field is taken
+as the ``solved'' outcome and no iterative, Krylov solver is applied
+to the pressure system. The outer solver of the mixed system uses the
+GCR algorithm. The algorithmic details are not discussed here, but a
+Krylov sub-space solver performs several global sums per iteration. As
+multiple solves per outer iteration are required, each itself of many
+iterations there is a reduction of several thousand global sums per
+time-step employing multigrid in ths manner.
```
Comment (samcusworth) [inline multigrid.tex]: "ths" -> "this"

File: multigrid.tex [hunk 4]
```diff
@@ -1 +1,93 @@
...
+
+Displayed in Table~\ref{tab:MG_data} and shown in figure~\ref{fig:mg}
+are the profiling data from CrayPAT for three different model versions
+of the Baroclinic wave test on a $C1152$ mesh with a $120$s time-step
+for $100$ time-steps. This corresponds to a roughly 9km
+resolution. The models were run on $384$ nodes with $6$ MPI ranks per
+node and $6$ Open MP threads per MPI rank.  The CSAR branch is
+\verb+r17264_CSAR19_gungho@17480+.
+
+Comparing the Krylov solver in the CSAR branch with the modification
```
Comment (samcusworth) [inline multigrid.tex:66]: Is it possible to extract the number of iterations from the log files?

File: scaling.tex [hunk 5]
```diff
@@ -1 +1,216 @@
...
+\centering\includegraphics[width=1.0\linewidth]{figs/wc-scale.eps}
+\caption{\label{fig:wc_scale}Wall-clock time strong scaling of the 
+  Baroclinic wave test. In red, labelled MPI is an MPI only
+  version. In yellow, labelled OMP is the hybrid MPI+OpenMP version.}
+\end{figure} 
+
+Shown in figure~\ref{fig:wc_scale} is the strong scaling of the whole
+model up to $384$ nodes. The MPI only version ran with $36$ MPI ranks
+per node. The hybrid version ran with $6$ MPI ranks per node and $6$
+openMP threads per MPI rank. The MPI only version on $384$ nodes
```
Comment (samcusworth) [inline scaling.tex]: "openMP" -> "OpenMP"

File: scaling.tex [hunk 6]
```diff
@@ -1 +1,216 @@
...
+crashed during the partitioning phase with an Out Of Memory (OOM)
+error. The global mesh is read in by every MPI rank and then
+partitioned. This can cause memory problems for large meshes on larger
+node counts. This can be solved by a single rank per node reading in
```
Comment (samcusworth) [inline scaling.tex:29]: "This can be solved"... does that mean we implemented the single rank per node reading?

File: scaling.tex [hunk 7]
```diff
@@ -1 +1,216 @@
...
+\begin{figure}[ht!]
+\centering\includegraphics[width=1.0\linewidth]{figs/comms-scale.eps}
+\caption{\label{fig:comms_scale}Wall-clock time strong scaling of the 
+  communications costs of the Baroclinic wave test. HE denotes Halo
+  Exchange and GS the Global Sum.}
+\end{figure} 
+
+The size of the halos is shown in Table~\ref{tab:scale-data}. The 
+halos for the hybrid version are bigger than the MPI only version, as 
+there are less MPI ranks, but there are less messages to send. As the
```
Comment (samcusworth) [inline scaling.tex]: Yes I will be that pedant...
"less" -> "fewer"

File: scaling.tex [hunk 8]
```diff
@@ -1 +1,216 @@
...
+  Exchange and GS the Global Sum.}
+\end{figure} 
+
+The size of the halos is shown in Table~\ref{tab:scale-data}. The 
+halos for the hybrid version are bigger than the MPI only version, as 
+there are less MPI ranks, but there are less messages to send. As the 
+hybrid version is running with six threads, it has six times less the 
+number of messages. The hybrid version has fewer, but bigger MPI 
+messages and moreover, will have less total data to send. It is not 
+obvious why it should be so much slower.
```
Comment (samcusworth) [inline scaling.tex:135]: Just a thought about the cause of the slowdown- have we reached a message size threshold? What do we set $MPICH_GNI_MAX_EAGER_MSG_SIZE to?

File: scaling.tex [hunk 9]
```diff
@@ -1 +1,216 @@
...
+the time-step is longer at $180$s and moreover, the FD data was not
+profiled using CrayPAT, but the time for $50$ time-steps calculated
+from the logs. Shown in table~\ref{fig:fd_comp} is a comparison of
+trunk at r17483 and the FD release. Clearly the hybrid version is now
+much slower. The trunk versions uses the new solver API, so the solver
+is a different algorithm but it is not doing a significantly different
+number of halo exchanges. Neither the YAXT library performing the halo
+exchanges nor the generated PSy layer code are any different, {\em
+  i.e.} halo exchanges called outside OpenMP regions, which call halo
+exchange on fields, which call YAXT. The FD model was not profiled
```
Comment (samcusworth) [inline scaling.tex]: We've already said that the FD release was not run with CrayPAT

**Comment by samcusworth** [comment]
I've read through the text and made a few comments- but not a thorough review.

File: multigrid.tex (hunk 2 above)
Comment (christophermaynard) [reply multigrid.tex:17]: 4 is not magic itself, but 4 tends work magically. 3 or less tends not to be enough, 5 tends to be too much, or at least no real gain but one also needs to be have at least 2**5=32 on the finest mesh per parallel MPI rank in order to have a minimum 1 horizontal degree of freedom per MPI rank on the coarsest mesh. Less than 1 requires messy reorganisation. Its usually an empirical measure.

Comment (christophermaynard) [reply multigrid.tex:17]: Other multigrid cycles are available. W and "full multigrid" which starts on the coarsest and goes up and down 1 level, then 2 and 1, etc. The idea being that spending more algorithm time on the coarsest levels where it is cheap to do the work. There are examples where this has been beneficial but often simplest is best and V-cycle is probably sufficient for us as we only solve to 10e-3 to 10e-6 anyway.

Comment (christophermaynard) [reply multigrid.tex:17]: Other multi grids are available. For example Algebraic. We are using Geometric multigrid. Our Prolongation and restriction operators use geometry (of the mesh). At a naive level MG just gets a cheaper, less good solution on a coarse mesh which starts the solution guess on the fine mesh. The real magic is the slow-to-converge, high-frequency modes on the fine mesh get aliased to low frequency modes on the coarse mesh (and vice versa) so the solution is much more optimal. Its possible to define Algebraic operators based on some kind of schur complement of the matrix one is trying to invert. This is highly non-trivial to do so most people resort to a library at this point. Hypre is a good one but only works with MPI decomposition (no shared memory) PETSc uses it, but it doesn't help if we are hybrid with Open MP or even a GPU implementation. Again, simple is probably best for us, especially as we gain 90% of the benefit with 10% of the work. Yes, V-cycle GMG is only 10% to of the work compared to some semi-exotic AMG thing. We do only solve until the time-step integration is stable, not to some highly accurate pressure field solution. Clear as mud?

File: multigrid.tex (hunk 1 above)
Comment (christophermaynard) [reply multigrid.tex]: Aah. It is no iterative solver, buy I guess my comma and Krylov rather muddies the sense of that sentence.
Revised to a simpler "no Krylov solver"

File: multigrid.tex (hunk 2 above)
Comment (samcusworth) [reply multigrid.tex:17]: Thanks for the extra details. I wrote a V-cycle MG solver back in my PhD days and also found that the optimum number of levels to use depended on the local size of the problem. I also got myself into trouble by asking my solver to solve too many levels - from 512**2 ending up with a 4x4 grid (yes we were all young a stupid once upon a time)- and it ruined the performance. In my experience having a minimum grid size - either explicit or implicit as we have here- is sensible.

File: scaling.tex [hunk 10]
```diff
@@ -1 +1,216 @@
...
+The Gung Ho dynamical core uses an unstructured mesh to avoid a
+singular pole which ultimately inhibits the scaling of the
+UM. Examining the scaling behaviour of the model to see if it
+performs sufficiently well is therefore important. To test the scaling a series
+of model runs was performed. The version of the LFRic repository trunk
+was $17483$. The code was compiled on the XCS with the Intel 17
+compiler, at production level which is \verb+-O3+. The baroclinic wave
+test was run on a $C576$ mesh. This roughly corresponds to a $17$km
+horizontal resolution. The model was run with a $30$km lid and with
+$30$ levels.  The time-step used was $180$s
```
Comment (samcusworth) [inline scaling.tex]: either "s" or "seconds" but not both

File: scaling.tex [hunk 11]
```diff
@@ -1 +1,216 @@
...
+the code is faster than the MPI version, whereas the opposite is now
+true for the data displayed here. It is unclear why this might be
+so. The science problem and parallel code are broadly the same. The
+major change is to the solver algorithm which uses the new solver
+API. However, it is far from obvious why this would change the
+behaviour.
+
+The jobs have been profiled with CrayPAT and the data is shown in
+table~\ref{tab:scale-data}. CrayPAT groups the data according certain
+categories. User code (labelled ``U'') in the table and is the LFRic
```
Comment (samcusworth) [inline scaling.tex]: Remove excess "and" here

File: scaling.tex [hunk 12]
```diff
@@ -1 +1,216 @@
...
+code, in this case kernels such as the various matrix-vector
+routines. The Halo Exchanges (labelled ``HE'') appear in the {\em ETC}
```
Comment (samcusworth) [inline scaling.tex]: the ETC category of the CrayPAT output? It's labelled as HE on the table

File: scaling.tex [hunk 13]
```diff
@@ -1 +1,216 @@
...
+shows in the profile at $1\%$ of run-time. The user code portion runs
+in roughly the same time as the $6$ thread version, which is faster
+than MPI only. However, the communication costs are much higher. Both
+Halo Exchange and Global Sum take much longer and have a high degree
+of imbalance, at around $30\%$ each. The number of Halo Exchanges is
+around 1.3 million, the same as MPI and $6$ thread versions.  It is
+reasonable to believe that at least some of this increase in cost is
+due to network variablility. This could be examined by running in
+large exclusive reservations or with a statistical analysis of a large
+number of runs. Both these are expensive in terms of resource use and
```
Comment (samcusworth) [inline scaling.tex]: "Both these" -> "Both of these"

File: scaling.tex [hunk 14]
```diff
@@ -1 +1,216 @@
...
+probably unjustifiable given pressure on resources for science rather
+than performance evaluation.
+
+It is also not obvious why the hybrid version should now be slower
+than MPI when the opposite was true in the FD release. In this data
+the time-step is longer at $180$s and moreover, the FD data was not
+profiled using CrayPAT, but the time for $50$ time-steps calculated
```
Comment (samcusworth) [inline scaling.tex]: "time for $50$ time-steps calculated" -> "time for $50$ time-steps can be calculated"

File: multigrid.tex [hunk 15]
```diff
@@ -1 +1,93 @@
...
+converging.
+This was ultimately resolved as an issue with the vertical
+pre-conditioner and not multigrid itself. 
+
+The multigrid algorithm implemented here is a four-level V-cycle of
+the horizontal mesh, and used to pre-condition the Helmholtz or
+pressure solver. In fact, the pre-conditioned pressure field is taken
+as the ``solved'' outcome and no Krylov solver is applied
+to the pressure system. The outer solver of the mixed system uses the
+GCR algorithm. The algorithmic details are not discussed here, but a
```
Comment (samcusworth) [inline multigrid.tex:22]: Am I the only one who doesn't know what the GCR algorithm is?

File: multigrid.tex [hunk 16]
```diff
@@ -1 +1,93 @@
...
+  CSAR branch with the Krylov subspace solver and ``Mg'' denotes
+  CSAR branch with the multigrid pre-conditioner as the solver.}
+\end{figure} 
+
+Displayed in Table~\ref{tab:MG_data} and shown in figure~\ref{fig:mg}
+are the profiling data from CrayPAT for three different model versions
+of the Baroclinic wave test on a $C1152$ mesh with a $120$s time-step
+for $100$ time-steps. This corresponds to a roughly 9km
+resolution. The models were run on $384$ nodes with $6$ MPI ranks per
+node and $6$ Open MP threads per MPI rank.  The CSAR branch is
```
Comment (samcusworth) [inline multigrid.tex]: "Open MP" -> "OpenMP"

File: multigrid.tex [hunk 17]
```diff
@@ -1 +1,93 @@
...
+pre-conditioner used as the solver has a further dramatic
+effect. Combined together they result in a speed up of $4 \times $ in
+wall clock time. Whilst this is a very welcome performance boost it is
+the effect on the communication costs which has greater benefit. The
+local communication of Halo exchange cost is reduced by roughly
+$3.5 \times $. The reduction in the cost of the global sums is very
+dramatic, more than $8 \times$. 
+
+The global sum is latency bound and so the cost increases with the
+number of MPI ranks taking part. Doing far less global sums will
```
Comment (samcusworth) [inline multigrid.tex]: less -> fewer

**Review by samcusworth — CHANGES_REQUESTED** [review]

File: multigrid.tex (hunk 3 above)
Comment (christophermaynard) [reply multigrid.tex]: Done.

File: multigrid.tex (hunk 4 above)
Comment (christophermaynard) [reply multigrid.tex:66]: I guess I could, I think I worked it at at 2500 less global sums per timestep!

Comment (christophermaynard) [reply multigrid.tex:66]: Oh, sorry. That was for multigrid. Not sure what you are asking now ...

File: scaling.tex (hunk 6 above)
Comment (christophermaynard) [reply scaling.tex:29]: No, it means we should though.  Added a footnote.

File: scaling.tex (hunk 7 above)
Comment (christophermaynard) [reply scaling.tex]: Not pedantry, sloppy = bad, Correct = good! fixed.

File: scaling.tex (hunk 8 above)
Comment (christophermaynard) [reply scaling.tex:135]: Yes, could be this. Not set to anything, so just the default. It was one of the things on my list to investigate. Do you know what the thresholds are?

File: scaling.tex (hunk 9 above)
Comment (christophermaynard) [reply scaling.tex]: Oops. Deleted second CrayPAT statement.

File: scaling.tex (hunk 10 above)
Comment (christophermaynard) [reply scaling.tex]: Oops. Deleted "s".

File: scaling.tex (hunk 11 above)
Comment (christophermaynard) [reply scaling.tex]: Done.

File: scaling.tex (hunk 12 above)
Comment (christophermaynard) [reply scaling.tex]: Yes, ETC for not user code or MPI. Meaning external library. I will clarify in the text.

File: scaling.tex (hunk 13 above)
Comment (christophermaynard) [reply scaling.tex]: Done.

File: scaling.tex (hunk 14 above)
Comment (christophermaynard) [reply scaling.tex]: Done.

File: multigrid.tex (hunk 16 above)
Comment (christophermaynard) [reply multigrid.tex]: Done.

File: multigrid.tex (hunk 17 above)
Comment (christophermaynard) [reply multigrid.tex]: Done.

**Comment by christophermaynard** [comment]
Committed changes to branch and pushed.

File: scaling.tex (hunk 8 above)
Comment (samcusworth) [reply scaling.tex:135]: I don't but I'll send you an email with the settings that we use in UM production runs.

File: scaling.tex [hunk 18]
```diff
@@ -1,4 +1,4 @@
-\section{\label{sec:scale}Scaling Analysis}
+6\section{\label{sec:scale}Scaling Analysis}
```
Comment (samcusworth) [inline scaling.tex:1]: Spurious 6?

**Comment by samcusworth** [comment]
Looks good to me... except a "6" character has appeared in scaling.tex. I'll remove this upon merge and commit

**Review by samcusworth — APPROVED** [review]

**Pull request merged**